### PR TITLE
deep-kernel-learning example, ported over from KernelFunctions.jl#322

### PR DESCRIPTION
**Pull request by st--**
Replaces https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/322

### CODE REVIEW

File: examples/2-deep-kernel-learning/script.jl [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+)
+Plots.scatter!(vcat(x_train...), y_train; label="data")
+pred = marginals(fp(x_test))
+Plots.plot!(vcat(x_test...), mean.(pred); ribbon=std.(pred), label="Prediction")
+
+# ## Training
+anim = Animation()
+nmax = 1000
+opt = Flux.ADAM(0.1)
+@showprogress for i in 1:nmax
```
Comment (st--) [inline examples/2-deep-kernel-learning/script.jl:61]: remove progress bar

File: examples/2-deep-kernel-learning/script.jl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    global grads = gradient(ps) do
+        loss(y_train)
+    end
+    Flux.Optimise.update!(opt, ps, grads)
+    if i % 100 == 0
+        L = loss(y_train)
+        @info "$i/$nmax; loss = $L"
+        p = Plots.plot(
+            vcat(x_test...), target_f; lab="true f", title="Loss = $(loss(y_train))"
```
Comment (st--) [inline examples/2-deep-kernel-learning/script.jl:70]: add iteration to label

**Review by willtebbutt — COMMENTED** [review]
A few things.

File: examples/2-deep-kernel-learning/script.jl [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# # Deep Kernel Learning with Flux
+# ## Package loading
```
[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:2]: edited inline

File: examples/2-deep-kernel-learning/script.jl [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# We use a couple of useful packages to plot and optimize
+# the different hyper-parameters
+using KernelFunctions
+using Flux
+using Distributions, LinearAlgebra
+using Plots
+using AbstractGPs
```
[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:9]: :roll_eyes: :joy:

[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:9]: btw could order-of-`using` ever have a functional effect in Julia?

File: examples/2-deep-kernel-learning/script.jl [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+# We create a simple 1D Problem with very different variations
+
+xmin, xmax = (-3, 3)  # Limits
+N = 150
+noise_std = 0.01
+x_train = collect(eachrow(rand(Uniform(xmin, xmax), N))) # Training dataset
+target_f(x) = sinc(abs(x)^abs(x)) # We use sinc with a highly varying value
+target_f(x::AbstractArray) = target_f(only(x))
+y_train = target_f.(x_train) + randn(N) * noise_std
+x_test = collect(eachrow(range(xmin, xmax; length=200))) # Testing dataset
```
[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:22]: I _think_ it's to do with Flux, but I'm not sure, that part I hadn't changed.
```suggestion
x_test_vec = range(xmin, xmax; length=200) # Testing dataset
x_test = collect(eachrow(x_test_vec)) # vector-of-vectors for Flux compatibility
```

File: examples/2-deep-kernel-learning/script.jl [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# ## Data creation
+# We create a simple 1D Problem with very different variations
+
+xmin, xmax = (-3, 3)  # Limits
+N = 150
+noise_std = 0.01
+x_train = collect(eachrow(rand(Uniform(xmin, xmax), N))) # Training dataset
+target_f(x) = sinc(abs(x)^abs(x)) # We use sinc with a highly varying value
+target_f(x::AbstractArray) = target_f(only(x))
+y_train = target_f.(x_train) + randn(N) * noise_std
```
Comment (st--) [inline examples/2-deep-kernel-learning/script.jl:21]:
```suggestion
y_train = target_f.(x_train_vec) + randn(N) * noise_std
```

File: examples/2-deep-kernel-learning/script.jl [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+x_test = collect(eachrow(range(xmin, xmax; length=200))) # Testing dataset
+
+plot(xmin:0.01:xmax, target_f; label="ground truth")
+scatter!(map(only, x_train), y_train; label="training data")
```
Comment (st--) [inline examples/2-deep-kernel-learning/script.jl:25]:
```suggestion
scatter!(x_train_vec, y_train; label="training data")
```

File: examples/2-deep-kernel-learning/script.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:22]: I guess it could be a ColVecs or something like that instead of a literal Vector[Vector[Float64]].

File: examples/2-deep-kernel-learning/script.jl [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+nmax = 1000
+opt = Flux.ADAM(0.1)
+for i in 1:nmax
+    global grads = gradient(ps) do
+        loss(y_train)
+    end
+    Flux.Optimise.update!(opt, ps, grads)
+    if i % 10 == 0
+        L = loss(y_train)
+        @info "$i/$nmax; loss = $L"
```
[the remark this answers is not in the document]
Comment (st--) [reply examples/2-deep-kernel-learning/script.jl:66]:
```suggestion
        if i % 100 == 0
            @info "$i/$nmax; loss = $L"
        end
```

File: examples/2-deep-kernel-learning/script.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (willtebbutt) [reply examples/2-deep-kernel-learning/script.jl]: It's certainly not meant to...

[the remark this answers is not in the document]
Comment (willtebbutt) [reply examples/2-deep-kernel-learning/script.jl]: But as far as following the style-guide goes, see https://github.com/invenia/BlueStyle#module-imports

**Review by willtebbutt — APPROVED** [review]
LGTM. Thanks for addressing the comments.
